quinn-proto/src/range_set/array_range_set.rs

```diff
@@ -0,0 +1,301 @@
+                    let next = self.0[idx + 1].clone();
+                    if curr.end >= next.start {
+                        self.0[idx].end = next.end.max(curr.end);
+                        self.0.remove(idx + 1);
+                    } else {
+                        break;
+                    }
+                }
+
+                return true;
+            }
+
+            idx += 1;
+        }
+
+        if x.start == x.end {
+            return result;
+        }
+
+        // Insert a range at the end
+        self.0.push(x);
+        true
+    }
+
+    pub fn remove(&mut self, mut x: Range<u64>) -> bool {
+        let mut result = false;
+
+        let mut idx = 0;
+        while idx != self.0.len() && x.start != x.end {
+            let range = self.0[idx].clone();
+
+            if x.end <= range.start {
+                // The range is fully before this range
+                return result;
+            } else if x.start >= range.end {
+                // The range is fully after this range
+                idx += 1;
+            } else {
+                // The range overlaps with this range
+                result = true;
+
+                // Trim everything off the range which is in front of this range
+                x.start = x.start.max(range.start);
+
+                // This is what we remove from this particular range
+                let to_remove = x.start..x.end.min(range.end);
+
+                // Everything which is after this range will be handled in the
+                // next loop iteration
+                x.start = range.end.min(x.end);
+
+                debug_assert!(range.start <= to_remove.start);
+                debug_assert!(to_remove.end <= range.end);
+
+                // We can end up with 2 parts after removal:
+                // - left is before the removed range
+                // - right is after the removed range
+                // Both ranges might be empty
+                let left = range.start..to_remove.start;
+                let right = to_remove.end..range.end;
+
+                let left_len = left.end - left.start;
+                let right_len = right.end - right.start;
+
+                debug_assert!(left_len + right_len <= range.end - range.start);
+
+                if left_len == 0 && right_len == 0 {
+                    // We drained the range
+                    self.0.remove(idx);
+                } else if left_len != 0 && right_len != 0 {
+                    // A left and right range exist
+                    // Use the existing slot for the right range, and insert
+                    // the left in front of it
+                    self.0[idx] = right;
+                    self.0.insert(idx, left);
+                    idx += 2;
+                } else if left_len != 0 {
+                    self.0[idx] = left;
+                    idx += 1;
+                } else {
+                    self.0[idx] = right;
+                    idx += 1;
+                }
```

Comment:
What's the motivation behind modifying `x` here? I feel like it might be a lot simpler without that. Something like:

```rust
let left = range.start .. x.start;
let right = x.end .. range.end;
if left.is_empty() && right.is_empty() {
    self.0.remove(idx)
} else if left.is_empty() {
    self.0[idx] = right;
    // etc
```

Reply:
I just found it easier to follow that way, and to figure out step by step what is kept and removed and have comments for everything. I can follow the `remove` method now rather well and have a high amount of confidence in it. The `insert` one is imho uglier.

Reply:
I still feel like the approach I outlined would be much simpler, as it avoids all arithmetic and reduces the number of intermediate values considerably.